### ISSUE TITLE
What is catalog's approach for validation against OSB API

### ISSUE THREAD

**Issue by derberg** [kubernetes-sigs/service-catalog#2088]
Hey, I have a generic inquiry to Service Catalog approach for validation of objects delivered by Service Brokers. 

Imagine situation. I register a service broker in catalog. Question is, who owns the validation?
* Do you expect the broker owner is responsible for validation against OSB API
* Are you making sure Service Catalog does it?
* Should I (as a cluster owner and person that installed catalog) make sure I put some layer on top of it or whatsoever?

We had a situation with one of the brokers. It had bug, and for schema for instance creation (https://github.com/openservicebrokerapi/servicebroker/blob/v2.13/spec.md#input-parameters-object), instead of returning object, we returned base64 string. Service Catalog did not reject broker registration, so we were able to create ServiceBroker resource and see classes with invalid schemas in the catalog.

Of course we fixed the bug as broker owners. I just don't know if it is also a bug on your side, that there were no errors propagated by catalog.

**Comment by derberg** [comment]
in case with base64 encoded schema I meant that instead of schema for instance creation like for example:

```
{
              "$schema": "http://json-schema.org/draft-04/schema#",
              "type": "object",
              "properties": {
                "billing-account": {
                  "description": "Billing account number used to charge use of shared fake server.",
                  "type": "string"
                }
              }
            }
```

our broker returned string in quotes:
`"ewogICAgICAgICAgICAgICIkc2NoZW1hIjogImh0dHA6Ly9qc29uLXNjaGVtYS5vcmcvZHJhZnQtMDQvc2NoZW1hIyIsCiAgICAgICAgICAgICAgInR5cGUiOiAib2JqZWN0IiwKICAgICAgICAgICAgICAicHJvcGVydGllcyI6IHsKICAgICAgICAgICAgICAgICJiaWxsaW5nLWFjY291bnQiOiB7CiAgICAgICAgICAgICAgICAgICJkZXNjcmlwdGlvbiI6ICJCaWxsaW5nIGFjY291bnQgbnVtYmVyIHVzZWQgdG8gY2hhcmdlIHVzZSBvZiBzaGFyZWQgZmFrZSBzZXJ2ZXIuIiwKICAgICAgICAgICAgICAgICAgInR5cGUiOiAic3RyaW5nIgogICAgICAgICAgICAgICAgfQogICAgICAgICAgICAgIH0KICAgICAgICAgICAgfQ=="`

so at the end this string was also returned in the `instanceCreateParameterSchema`
---

other fresh example is that in new broker we return empty schema so ServiceClass info is `"instanceCreateParameterSchema": ""` and to my understanding of https://github.com/openservicebrokerapi/servicebroker/blob/v2.13/spec.md#input-parameters-object it must also be a valid Json Schema object, and if you want to return empty schema it still has to be an empty Json Schema object and not just empty attribute

---

these are the use cases where we are not sure if these are bugs on Service Catalog side or as I wrote in the intro, maybe this intentional.

**Comment by jberkhahn** [comment]
I'm still a bit confused here. The instance creation scheme is supposed to be defined as an array of JSON objects like so:
```
"schemas": {
        "service_instance": {
          "create": {
            "parameters": {
              "$schema": "http://json-schema.org/draft-04/schema#",
              "type": "object",
              "properties": {
                "billing-account": {
                  "description": "Billing account number used to charge use of shared fake server.",
                  "type": "string"
                }
              }
            }
          },
          "update": {
            "parameters": {
              "$schema": "http://json-schema.org/draft-04/schema#",
              "type": "object",
              "properties": {
                "billing-account": {
                  "description": "Billing account number used to charge use of shared fake server.",
                  "type": "string"
                }
              }
            }
          }
        },
        "service_binding": {
          "create": {
            "parameters": {
              "$schema": "http://json-schema.org/draft-04/schema#",
              "type": "object",
              "properties": {
                "billing-account": {
                  "description": "Billing account number used to charge use of shared fake server.",
                  "type": "string"
                }
              }
            }
          }
        }
```
So, it's an an array of objects under a couple sub headings. There is no attribute called "instanceCreateParameterSchema". Are you sure your catalog response is structured correctly? If you have the field named/structured incorrectly, Service Catalog probably ignores the incorrect fields entirely and parses the catalog as if they didn't exist. You can take a look at where the schemas actually get parsed [here.](https://github.com/kubernetes-incubator/service-catalog/blob/master/pkg/controller/controller.go#L1206) It looks like we attempt to parse it as a JSON, and if its parseable it just gets thrown in.

**Comment by derberg** [comment]
sorry, I think it is because of the shortcuts I'm taking here, once I talk about Service Class in ServiceBroker, sometimes about Service Class in the Service Catalog.

When I write about `instanceCreateParameterSchema` I mean Service Catalog side of the Service Class: https://github.com/kubernetes-incubator/service-catalog/blob/master/pkg/apis/servicecatalog/v1beta1/types.go#L614
which in case of the Service Class from Service Broker perspective is:
```
"service_instance": {
          "create": {
            "parameters": {
              "$schema": "http://json-schema.org/draft-04/schema#",
              "type": "object",
              "properties": {
                "billing-account": {
                  "description": "Billing account number used to charge use of shared fake server.",
                  "type": "string"
                }
              }
            }
          }
```

so in one case `instanceCreateParameterSchema` returned empty string because the broker returned empty `service_instance.create.parameters` object. @mszostok correct me if I'm wrong here

so basically according to the OSB API, the broker should never return empty string in `service_instance.create.parameters`. If no parameters, it should return empty schema.

**Comment by jberkhahn** [comment]
Ok, discussed this with the OSB PMC. We're of the opinion that empty parameters should be allowed, as it means the thing supports no parameters. 

need to fix the bug on our end and actually validate the schema rather than just parsing it, though

**Comment by derberg** [comment]
great, just to be 100% sure we understand each other fully: but `empty parameters should be allowed` 
you do not mean that empty string is allowed right? but rather something like below examples?:

```
{ 
   "description": "voice mail record", 
   "type": "object", 
   "additionalProperties": false, 
   "properties": { 
    } 
  } 
}
```
or
```
{
            "$schema": "http://json-schema.org/draft-04/schema#",
            "additionalProperties": false,
            "type": "object"
}
```

**Comment by jberkhahn** [comment]
sorry, was a bit incorrect in my previous statement. it should be
```
{
            "$schema": "http://json-schema.org/draft-04/schema#",
            "additionalProperties": false,
}
```
